Management of in-mesh certificates (#13344)
istio/istio.io: content/en/docs/ops/configuration/traffic-management/manage-mesh-certificates/index.md

your Gateway see the the [Secure Gateways](/docs/tasks/traffic-management/ingress/secure-ingress/) document. For managing the CA certificated istiod uses to generate workload certificates for see
the [Plugin CA Certificates](/docs/tasks/security/cert-management/plugin-ca-cert/) document.

It is also recommended that users seek to use a more robust tool to manage and issue certificates
using [cert-manager](/docs/ops/integrations/certmanager/) or a cloud-vendor based managed
offering.

## istiod

When Istio is installed without a root CA certificate plugged-in, istiod will mint a self-signed
CA certificate using RSA 2048.

Istiod will also create a certificate for the istiod service derived from this root CA, whether
it was created by istiod or plugged-in.

Changing this will require a istiod modifying either the IOP file you give to
istioctl or the values file you pass into your Helm install for the istio-discovery chart.

{{< tip >}}
While there are many environment variables that can be changed for
[pilot-discovery](/docs/reference/commands/pilot-discovery/) this document will only
outline some of them.
{{< /tip >}}

> {{< tabset category-name="gateway-install-type" >}}

Author: ```suggestion
{{< tabset category-name="certificates" >}}
```

{{< /text >}}

{{< /tab >}}

{{< tab name="Helm" category-value="helm" >}}

{{< text yaml >}}
pilot:
  env:
    CITADEL_SELF_SIGNED_CA_RSA_KEY_SIZE: 4096
    <key>: <value>
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

### Self-signed CA

To control the type of signature algorithm used by istiod you must specify
`CITADEL_SELF_SIGNED_CA_CERT_ALGORITHM` to be either `RSA` (the default, does not have to be set)
or `ECC`.

Managing other properties of this certificate based on its type can be managed by setting
`CITADEL_SELF_SIGNED_CA_RSA_KEY_SIZE`, `ECC_CURVE` and `ECC_SIGNATURE_ALGORITHM`.

### istiod Certificate

To control the type of certificate generated by istiod for itself is only done by specifying
`ECC_SIGNATURE_ALGORITHM` to `ECDSA`.

It's strength can be configured either with `ECC_CURVE` or `WORKLOAD_RSA_KEY_SIZE`.

## Sidecars

Since sidecars manage their own certificates for in-mesh communication, the sidecars
are responsible for managing their private keys and generated CSRs. The sidecar
injector needs to be modified to inject the environment variables to be used for
this purpose.

{{< tip >}}
While there are many environment variables that can be changed for

> [pilot-agent](/docs/reference/commands/pilot-agent/) this document will only

Reviewer: ```suggestion
[pilot-agent](/docs/reference/commands/pilot-agent/), this document will only
```

outline some of them.
{{< /tip >}}

{{< tabset category-name="gateway-install-type" >}}

{{< tab name="IstioOperator" category-value="iop" >}}

{{< text yaml >}}
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    defaultConfig:
      proxyMetadata:
        CITADEL_SELF_SIGNED_CA_RSA_KEY_SIZE: 4096
{{< /text >}}

{{< /tab >}}

{{< tab name="Helm" category-value="helm" >}}

{{< text yaml >}}
meshConfig:
  defaultConfig:
    proxyMetadata:
      CITADEL_SELF_SIGNED_CA_RSA_KEY_SIZE: 4096
{{< /text >}}

{{< /tab >}}

{{< tab name="Annotation" category-value="annotation" >}}

{{< text yaml >}}
apiVersion: apps/v1
kind: Deployment
metadata:
  name: sleep
spec:
  ...
  template:
    metadata:
      ...
      annotations:
        ...
        proxy.istio.io/config: |
          CITADEL_SELF_SIGNED_CA_RSA_KEY_SIZE: 4096
    spec:
      ...
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

### Signature Algorithm

> By default, the sidecars will create RSA based certificates. If the user wishes to change it to

Reviewer: ```suggestion
By default, the sidecars will create RSA-based certificates. If you want to change it to
```

> ECC then `ECC_SIGNATURE_ALGORITHM` must be set to `ECDSA`

Reviewer: ```suggestion
ECC, you need to set `ECC_SIGNATURE_ALGORITHM` to `ECDSA`.
```

{{< tabset category-name="gateway-install-type" >}}

{{< tab name="IstioOperator" category-value="iop" >}}

{{< text yaml >}}
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    defaultConfig:
      proxyMetadata:
        ECC_SIGNATURE_ALGORITHM: "ECDSA"
{{< /text >}}

{{< /tab >}}

{{< tab name="Helm" category-value="helm" >}}

{{< text yaml >}}
meshConfig:
  defaultConfig:
    proxyMetadata:
      ECC_SIGNATURE_ALGORITHM: "ECDSA"
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

Only P256 and P384 are supported via `ECC_CURVE`.

> To modify the RSA key size if you wish to continue using RSA signature algorithms

Reviewer: ```suggestion
If you prefer to retain RSA signature algorithms and want to modify the RSA key size,
```

> you can change `WORKLOAD_RSA_KEY_SIZE`.

Reviewer: ```suggestion
you can change the value of `WORKLOAD_RSA_KEY_SIZE`.
```

injector needs to be modified to inject the environment variables to be used for
this purpose.

{{< tip >}}
While there are many environment variables that can be changed for
[pilot-agent](/docs/reference/commands/pilot-agent/), this document will only
outline some of them.
{{< /tip >}}

{{< tabset category-name="gateway-install-type" >}}

{{< tab name="IstioOperator" category-value="iop" >}}

{{< text yaml >}}
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    defaultConfig:
      proxyMetadata:
        CITADEL_SELF_SIGNED_CA_RSA_KEY_SIZE: 4096
{{< /text >}}

{{< /tab >}}

{{< tab name="Helm" category-value="helm" >}}

{{< text yaml >}}
meshConfig:
  defaultConfig:
    proxyMetadata:
      CITADEL_SELF_SIGNED_CA_RSA_KEY_SIZE: 4096
{{< /text >}}

{{< /tab >}}

{{< tab name="Annotation" category-value="annotation" >}}

{{< text yaml >}}
apiVersion: apps/v1
kind: Deployment
metadata:
  name: sleep
spec:
  ...
  template:
    metadata:
      ...
      annotations:
        ...
        proxy.istio.io/config: |
          CITADEL_SELF_SIGNED_CA_RSA_KEY_SIZE: 4096
    spec:
      ...
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

### Signature Algorithm

> By default, the sidecars will create RSA-based certificates. If you want to change it to

Reviewer 2: I think this line needs `RSA-based` added to `.spelling`, and that just makes me angry.
